addons/frag/functions/fnc_fired.sqf

```diff
@@ -19,40 +19,71 @@
-    private _skip = getNumber (configFile >> "CfgAmmo" >> _ammo >> QGVAR(skip));
-    private _explosive = getNumber (configFile >> "CfgAmmo" >> _ammo >> "explosive");
-    private _indirectRange = getNumber (configFile >> "CfgAmmo" >> _ammo >> "indirectHitRange");
-    private _force = getNumber (configFile >> "CfgAmmo" >> _ammo >> QGVAR(force));
-    private _fragPower = getNumber (configFile >> "CfgAmmo" >> _ammo >> "indirecthit") * (sqrt (getNumber (configFile >> "CfgAmmo" >> _ammo >> "indirectHitRange")));
+if (_ammo isEqualTo "" || {isNull _projectile} ||
+    {_projectile getVariable [QGVAR(blacklisted), false]}) exitWith {
+    TRACE_2("bad ammo or projectile, or blackList",_ammo,_projectile);
+};
 
-    _shouldAdd = (_skip == 0) && {(_force == 1) || {_explosive > 0.5 && {_indirectRange >= 4.5} && {_fragPower >= 35}}};
+if (GVAR(spallEnabled) && {_ammo call FUNC(shouldSpall)}) then {
+    private _hitPartEventHandler = _projectile addEventHandler [
+        "HitPart",
+        {
+            params ["_projectile", "_hitObject", "", "_posASL", "_velocity"];
 
-    if (GVAR(spallEnabled) && {!_shouldAdd}) then {
-        private _caliber = getNumber (configFile >> "CfgAmmo" >> _ammo >> "caliber");
-        if !(_caliber >= 2.5 || {(_explosive > 0 && {_indirectRange >= 1})}) exitWith {}; // from check in doSpall: line 34
-        TRACE_1("Won't frag, but will spall",_caliber);
-        _shouldAdd = true;
-    };
+            // starting v2.18 it may be faster to use the instigator EH parameter, the same as the second entry shotParents, to recreate _shotParent
+            // The "explode" EH does not get the same parameter
+            private _shotParents = getShotParents _projectile;
+            private _ammo = typeOf _projectile;
 
-    TRACE_6("Setting Cache",_skip,_explosive,_indirectRange,_force,_fragPower,_shouldAdd);
-    GVAR(cacheRoundsTypesToTrack) set [_ammo, _shouldAdd];
+            /*
+             * Wait a frame to see what happens to the round, may result in
+             * multiple hits / slowdowns getting shunted to the first hit
+             */
+            [
+                // only let a unit make a frag event once per ACE_FRAG_SPALL_UNIT_HOLDOFF
+                {
+                    private _shotParents = _this#5;
+                    if (CBA_missionTime < _shotParents#1 getVariable [QGVAR(nextSpallEvent), -1]) exitWith {};
+                    _this call FUNC(doSpall);
+                },
+                [_hitObject, _ammo, _projectile, _posASL, _velocity, [objNull, _shotParents#1]]
+            ] call CBA_fnc_execNextFrame;
+        }
+    ];
+    _projectile setVariable [QGVAR(hitPartEventHandler), _hitPartEventHandler];
 };
 
-if (_shouldAdd) then {
-    // firedMan will have nil "_gunner", so just check _unit; for firedVehicle we want to check _gunner
-    private _localShooter = if (isNil "_gunner") then {local _unit} else {local _gunner};
```

Comment:
we still need some kind of locality check like this
`ace_firedNonPlayer` is going to trigger on all machines for AI

e.g. put down AI and make him throw a grenade and frago will trigger multiple times

Reply:
I didn't see that, or at least notice it. I'll have something in an hour here.

Reply:
Simplest approach is still best I guess. For reference using `local ((getShotParents _projectile)#1)` is slightly more efficient, but struggles with Zeus controlled units.